### ISSUE TITLE
Unable to bind to previously bound port after pod restart (Operation not permitted)

### ISSUE THREAD

**Issue by rastislavs** [cilium/cilium#19067]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

An envoy pod binds to several TCP ports, e.g. `0.0.0.0:30539` to act as a tcp proxy. After a pod restart, it is unable to bind to the ports used for the tcp proxy again with the following error:
```
cannot bind '0.0.0.0:30539': Operation not permitted
```
This issue only appears in clusters with Cilium CNI with ebpf kube-proxy-replacement enabled. With other CNIs or with Cilium with iptables/IPVS kube-proxy mode enabled the issue does not appear in the same environment.

When the issue happens, the netstat output in the affected pod does not show that the affected port (`30539`in this case) would be used:
```
/ # netstat -tulpn
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 0.0.0.0:8002            0.0.0.0:*               LISTEN      1/envoy
tcp        0      0 127.0.0.1:9001          0.0.0.0:*               LISTEN      1/envoy
tcp        0      0 :::8001                 :::*                    LISTEN      -
tcp        0      0 :::8080                 :::*                    LISTEN      -
/ #
```
Still, binding to the affected port does not work for any application, which can be reproduced e.g. with netcat:
```
/ # nc -l -p 30539 -s 0.0.0.0
nc: bind: Operation not permitted
/ #
```
Binding to a different port works fine:
```
/ # nc -l -p 30540 -s 0.0.0.0
^Cpunt!

/ #
```
Binding to the affected (non-working) port, but with a different local address (pod IP) works fine:
```
/ # nc -l -p 30539 -s 172.25.0.145
^Cpunt!

/ #
```

Envoy config dump: [envoy_config_dump.txt](https://github.com/cilium/cilium/files/8198489/envoy_config_dump.txt)
 

### Cilium Version

v1.11.0

### Kernel Version

Linux 5.4.0-100-generic #113-Ubuntu SMP Thu Feb 3 18:43:29 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

v1.21.8

### Sysdump

Because of a large file, I split it into multiple archives:
- [cilium-sysdump-20220307-145616.zip](https://github.com/cilium/cilium/files/8198320/cilium-sysdump-20220307-145616.zip)
- [cilium-bugtool-cilium-7fdbm-20220307-145616.zip](https://github.com/cilium/cilium/files/8198327/cilium-bugtool-cilium-7fdbm-20220307-145616.zip)
- [cilium-bugtool-cilium-hktpb-20220307-145616.zip](https://github.com/cilium/cilium/files/8198330/cilium-bugtool-cilium-hktpb-20220307-145616.zip)
- [cilium-bugtool-cilium-jfsxd-20220307-145616.zip](https://github.com/cilium/cilium/files/8198335/cilium-bugtool-cilium-jfsxd-20220307-145616.zip)

 

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by rastislavs** [comment]
Hi @gandro,
thanks for the reply.

> With kube-proxy-replacement enabled, we're blocking bind calls to any ports used by NodePort in the host namespace, as NodePort BPF will hijack any connections to that port.

The affected pod is not running in the host network namespace, it is a standard pod in its own network namespace.

> I am surprised to hear that you don't see this with regular kube-proxy however.

Yes, this has been working for us for years with kube-proxy. It also works with Cilium + kube-proxy-replacement, but only until the affected pod restarts. But as I mentioned, this is not a host network pod.

**Comment by gandro** [comment]
Thanks. It shouldn't block non-host network pods, that does sound potentially like a bug.

/cc @brb

**Comment by gandro** [comment]
Ah, I think I understand what's going on: The necessary kernel features to determine the network namespace of the `bind` call was only added in Linux 5.7 ([commit](https://lore.kernel.org/bpf/cover.1585323121.git.daniel@iogearbox.net/T/))

Thus, in your setup with kernel 5.4 `kube-proxy-replacement` is not able to distinguish between the pod and the host namespace and blocks basically all NodePort ports, regardless of the namespace. You can disable the bind protection via `node-port-bind-protection` CLI flag or ConfigMap option - which will allow processes to bind to the NodePort from the host network namespace, which is not really what you want either.

**Comment by rastislavs** [comment]
Thanks @gandro, I can confirm that the issue does not occur on a newer kernel version (tried with 5.13).

What is not clear to me is why the bind to a port from the NodePort range initially works, and only after the pod restart it stops working for the port that was previously working. And the bind still works for a different NodePort (not yet used).

**Comment by gandro** [comment]
> What is not clear to me is why the bind to a port from the NodePort range initially works, and only after the pod restart it stops working for the port that was previously working. And the bind still works for a different NodePort (not yet used).

Cilium only blocks `bind` calls for ports actually in use, that explains the latter.

We would need more details on your pod restart sequence, but I assume it's a race. Generally speaking Cilium can only block new `bind` calls. If the socket is already bound when Cilium is started or restarted, then it won't touch already bound sockets.
